### DISCUSSION TITLE
Control of shadow casting

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3839] [Q&A] [answered]
Is there any way to make a geometry cast shadows from certain lights and not from other generally shadow-casting lights?  I'm guessing the answer is No, but figured I might as well ask.  If you wonder why I would want such a thing, I want to make an invisible mask to put in front of a spot light to change the shape of the beam.  And the reason I want to do THAT is to see if I can fake a shadow-casting point light by putting together 4 spot lights.

**Comment by romainguy** [comment, marked as answer]
This is something we've been thinking about but never prioritize. It's basically the idea of light channels (not just for shadows but also for lighting in general). It would actually be fairly easy and cheap to implement.

@prideout @bejado

**Comment by TheOriginalBytePlayer** [comment]
I would second (**and third, and fourth...**) this request, as it is a show stopper for us trying to move from OpenGL to filament.  We make a software program that is, in effect, a virtual film studio, and having flexible lighting capabilities with multiple lights is vital to us.   Do you guys do bounties or any such thing for us to bump this up in the priority list, and/or would you have guidance in how we might be able to contribute this feature ourselves (given it being "cheap and easy to implement")?

**Reply by romainguy** [reply]
This should be fairly simple to implement, each light needs to be assigned a bitmask (just a byte probably) and each renderable would need the same.

**Reply by TheOriginalBytePlayer** [reply]
Any further thoughts on making this happen -- it would be a godsend for us...

**Reply by TheOriginalBytePlayer** [reply]
@romainguy @jwwalker I noticed that in the v1.12.0 release in August 2021 you added these light channels.  What is the impact of this in terms of number of lights, both shadow-casting and otherwise?

**Reply by romainguy** [reply]
There's no impact.

**Reply by TheOriginalBytePlayer** [reply]
Well, that's disappointing.  Could you tell me what the current story is on how many lights you can have in a scene, what kind, which of them do shadow casting and how many of those you can have.

Also, would it be possible to dynamically switch the shadow casting to just the lights in the frustum or the closest ones, or something of that order.

Lastly are there any plans to increase this number, even accepting a significant slowdown, so that one might have a fast low-shadow mode and a slower full shadow mode?  Thanks again for all your help...

**Reply by romainguy** [reply]
@bejado How do we currently select which lights cast shadows when there are too many requesting it?

**Reply by bejado** [reply]
- You can have a single directional light; it can either cast shadows or not.
- You can have [256 total punctual lights](https://github.com/google/filament/blob/fbfcce563d64357138cd29d3c65698f8bd64c06b/libs/filabridge/include/private/filament/EngineEnums.h#L60) (point and spot).
- Point lights cannot cast shadows.
- At most [14 spotlights can cast shadows](https://github.com/google/filament/blob/fbfcce563d64357138cd29d3c65698f8bd64c06b/libs/filabridge/include/private/filament/EngineEnums.h#L63-L66).

We select the first 14 spotlights that are [visible in the frustum](https://github.com/google/filament/blob/433c163c61799934407e9cf8dcbe9f39974e84e6/filament/src/details/View.cpp#L795), [sorted by distance to the camera](https://github.com/google/filament/blob/433c163c61799934407e9cf8dcbe9f39974e84e6/filament/src/details/View.cpp#L827).
Note you have total control over which lights cast shadows via `LightManager::setShadowCaster`.

**Reply by romainguy** [reply]
You can also control which objects receive shadows, and with light channels, which objects are affected by which lights.

**Reply by pixelflinger** [reply]
A few more details: lights that don't impact the view are culled first (e.g. if their influence doesn't intersect with the view frustum) and you can have a total of 256 *active* lights. These lights are sorted w.r.t. the camera, so if there are too many shadows requested, we select the ones cast by lights closer to the camera (note that a light close to the camera can cast a shadow far from the camera, so this is not a perfect solution, but it's well defined).

**Comment by TheOriginalBytePlayer** [comment]
Thank you for all the detail.  Unfortunately without point light shadows we still can’t use filament in our program as it’s a virtual film studio and shadows are a big part of the filmmaking process.  Is this on the roadmap or is it just not going to happen?  Similarly, are there any plans for area lights?

 

Final (for now!) stupid question, vis-a-vis the lights/shadows.   Would there possibly be a multi-pass solution where you would render the whole scene once then render just the missing shadows on top of it?  I know that would be a performance hit but it would allow us to have a quick mode and then a detail mode for printing, final rendering.

 

Thanks in advance for your time, and for making such an amazing render freely available.  I REALLY want to use it, but at the moment, the lighting/shadow situation is probably a show stopper for us… but it’s SO close.  

 

Cheers,

 

Ken Schafer

President / Lead Program Architect
Innoventive Software

Makers of the Emmy Award winning

Virtual film studio, FrameForge Studio

www.frameforge.com

**Reply by romainguy** [reply]
We'd like to add both shadows for point lights, and area lights.

A multi-pass solution would unfortunately not work given how our rendering is currently architected. For your particular case maybe we could start with adding support for point light shadows on desktop only? @pixelflinger @bejado wdyt?

**Reply by pixelflinger** [reply]
yes, there's a few non-trivial steps:
- ShadowMapManager will have to manage another texture with cubemap arrays
- generating the shadowmap itself is tricky. We could do the 6 face separately, but that's not very efficient. Usually it's done with the geometry shader to output all 6 faces at once. But we don't have support for that.

It's all doable, but it's a fair amount of work.

**Reply by bejado** [reply]
@pixelflinger We could limit point shadow support for Feature level 2 and above. We could also hold off on adding geometry shader support. It sounds like the performance requirements are somewhat flexible. Still a good bit of work though.

**Reply by pixelflinger** [reply]
I had an idea this weekend... stay tuned :-) [no promises!]

**Reply by pixelflinger** [reply]
@TheOriginalBytePlayer @jwwalker If someone wants to try #6152 and see how that works...

**Comment by TheOriginalBytePlayer** [comment]
We only run on desktop, so that would make all the difference…!

**Comment by TheOriginalBytePlayer** [comment]
We are in the midst of development on our next version – which is the one we hope to use filament on – and it’s at least four months away from beta testing, so there is time, if that helps at all with your decision.  I think this would be of a lot of use to a lot of your users so I do hope you strongly consider doing it…!

 

From: ***@***.*** ***@***.***> 
Sent: Friday, September 30, 2022 1:36 PM
To: 'google/filament' ***@***.***>; 'google/filament' ***@***.***>
Cc: 'TheOriginalBytePlayer' ***@***.***>; 'Comment' ***@***.***>
Subject: RE: [google/filament] Control of shadow casting (Discussion #3839)

 

We only run on desktop, so that would make all the difference…!

**Reply by romainguy** [reply]
See PR #6152 😄

**Comment by TheOriginalBytePlayer** [comment]
I await with baited breath (which does NOT mean I feel sorry for myself and are going to go eat worms… though I admit I’m hard pressed to come up with another good meaning! :) )

**Reply by jwwalker** [reply]
> I await with baited breath 

I can't resist being pedantic here.  The term is "bated breath", as in "I'm holding my breath", not "I made my breath smell like tuna so as to attract cats".
